### PR TITLE
Ensure TempDirectory can return a File

### LINKED ISSUE #1725: Support File parameters in TempDirectory extension
## Overview

The `TempDirectory` extension is great but only supports `Path`. I think it does not cost much to support `File`, and it enables to get a more fluent code base for several cases where the `Path` API is not yet the first citizen.

## Deliverables

- [x] Support `File` in addition to `Path` in the `TempDirectory`
- [x] Update Javadoc for `TempDirectory`
- [x] Update "The TempDirectory Extension" section of the User Guide
- [x] Document change in the Release Notes

### PR DESCRIPTION
**Pull request by rmannibucau**
Resolves #1725.

### CODE REVIEW

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/support/io/TempDirectory.java [hunk 1]
```diff
@@ -220,15 +221,26 @@ public boolean supportsParameter(ParameterContext parameterContext, ExtensionCon
...
 					CloseablePath.class) //
 				.get();
+		if (expectsPath) {
+			return path;
+		}
+		try {
+			return path.toFile();
+		}
+		catch (final UnsupportedOperationException uoe) { // not default filesystem
+			throw new ParameterResolutionException("Unsupported FileSystem in tempDirProvider: " + tempDirProvider,
```
Comment (sbrannen) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/support/io/TempDirectory.java]: Please introduce a test for this exception handling.

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/support/io/TempDirectory.java [hunk 2]
```diff
@@ -220,15 +221,25 @@ public boolean supportsParameter(ParameterContext parameterContext, ExtensionCon
...
 					key -> tempDirProvider.get(parameterContext, extensionContext, TEMP_DIR_PREFIX),
 					CloseablePath.class) //
 				.get();
+		if (expectsPath) {
+			return path;
+		}
+		try {
+			return path.toFile();
+		}
+		catch (final UnsupportedOperationException uoe) { // not default filesystem
```
[the remark this answers is not in the document]
Comment (marcphilipp) [reply junit-jupiter-api/src/main/java/org/junit/jupiter/api/support/io/TempDirectory.java]: this one, too. 😉

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/support/io/TempDirectory.java [hunk 3]
```diff
@@ -45,7 +46,8 @@
  *
  * <p>The temporary directory is only created if a test or lifecycle method or
  * test class constructor has a parameter annotated with
- * {@link TempDir @TempDir}. If the parameter type is not {@link Path} or if the
+ * {@link TempDir @TempDir}. If the parameter type is
+ * neither {@link Path} not {@link File} or if the
```
Comment (sbrannen) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/support/io/TempDirectory.java]: not --> nor

Comment (marcphilipp) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/support/io/TempDirectory.java]:
```suggestion
 * neither {@link Path} nor {@link File} or if the
```

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/support/io/TempDirectory.java [hunk 4]
```diff
@@ -220,15 +222,26 @@ public boolean supportsParameter(ParameterContext parameterContext, ExtensionCon
...
 					key -> tempDirProvider.get(parameterContext, extensionContext, TEMP_DIR_PREFIX),
 					CloseablePath.class) //
 				.get();
+		if (expectsPath) {
+			return path;
+		}
+		try {
+			return path.toFile();
+		}
+		catch (final UnsupportedOperationException uoe) { // not default filesystem
```
Comment (sbrannen) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/support/io/TempDirectory.java]:
```suggestion
		catch (UnsupportedOperationException uoe) { // not default filesystem
```

File: junit-jupiter-engine/src/test/java/org/junit/jupiter/api/support/io/TempDirectoryTests.java [hunk 5]
```diff
@@ -340,11 +341,24 @@ void check(@TempDir Path tempDir) {
...
 	}
 
+	@Nested
+	@DisplayName("User can rely on java.io.File injection type")
+	@ExtendWith(TempDirectory.class)
+	class FileInjection {
+		@Test
+		@DisplayName("File and Path injections lead to the same folder/behavior")
+		void checkFile(@TempDir File tempDir, @TempDir Path ref) {
+			final Path path = tempDir.toPath();
```
Comment (sbrannen) [inline junit-jupiter-engine/src/test/java/org/junit/jupiter/api/support/io/TempDirectoryTests.java]:
```suggestion
			Path path = tempDir.toPath();
```

File: junit-jupiter-api/src/main/java/org/junit/jupiter/api/support/io/TempDirectory.java [hunk 6]
```diff
@@ -220,15 +222,26 @@ public boolean supportsParameter(ParameterContext parameterContext, ExtensionCon
...
+			throw new ParameterResolutionException("Unsupported FileSystem in tempDirProvider: " + tempDirProvider,
```
Comment (sbrannen) [inline junit-jupiter-api/src/main/java/org/junit/jupiter/api/support/io/TempDirectory.java]: Please introduce a test for the "unsupported filesystem" use case.

**Comment by sbrannen** [comment]
@rmannibucau, I just posted a new review requesting a few minor changes and an additional test method.

Once you've done that, I think this PR will be ready to be merged into `master`.

**Comment by rmannibucau** [comment]
@sbrannen accepted the changes and added the test

**Comment by sbrannen** [comment]
@marcphilipp, I think this PR is ready. 

PTAL and either dismiss or accept the changes as appropriate.

**Comment by sbrannen** [comment]
> @sbrannen accepted the changes and added the test

Thanks

**Comment by sbrannen** [comment]
This has been merged into `master` in e8140c0a9f83fe9427576642e822df9dd2acd44a and polished in 2bc09b99e29ae7727ab42591564c393e0f5ff284.

Thanks for the contribution!

**Pull request merged**